Mathlib/AlgebraicGeometry/Morphisms/QuasiCompact.lean

```diff
@@ -0,0 +1,356 @@
+    MorphismProperty.StableUnderComposition @QuasiCompact := fun _ _ _ _ _ _ _ => inferInstance
+#align algebraic_geometry.quasi_compact_stable_under_composition AlgebraicGeometry.quasiCompact_stableUnderComposition
+
+--attribute [-simp] PresheafedSpace.as_coe SheafedSpace.as_coe
+
+theorem QuasiCompact.affineProperty_stableUnderBaseChange :
+    QuasiCompact.affineProperty.StableUnderBaseChange := by
+  intro X Y S _ _ f g h
+  rw [QuasiCompact.affineProperty] at h ⊢
+  skip
+  let 𝒰 := Scheme.Pullback.openCoverOfRight Y.affineCover.finiteSubcover f g
+  have : Finite 𝒰.J := by dsimp; infer_instance
+  have : ∀ i, CompactSpace (𝒰.obj i).carrier := by intro i; dsimp; infer_instance
+  exact 𝒰.compactSpace
+#align algebraic_geometry.quasi_compact.affine_property_stable_under_base_change AlgebraicGeometry.QuasiCompact.affineProperty_stableUnderBaseChange
+
+theorem quasiCompact_stableUnderBaseChange : MorphismProperty.StableUnderBaseChange @QuasiCompact :=
+  quasiCompact_eq_affineProperty.symm ▸
+    QuasiCompact.affineProperty_isLocal.stableUnderBaseChange
+      QuasiCompact.affineProperty_stableUnderBaseChange
+#align algebraic_geometry.quasi_compact_stable_under_base_change AlgebraicGeometry.quasiCompact_stableUnderBaseChange
+
+variable {Z : Scheme.{u}}
+
+instance (f : X ⟶ Z) (g : Y ⟶ Z) [QuasiCompact g] :
+    QuasiCompact (pullback.fst : pullback f g ⟶ X) :=
+  quasiCompact_stableUnderBaseChange.fst f g inferInstance
+
+instance (f : X ⟶ Z) (g : Y ⟶ Z) [QuasiCompact f] :
+    QuasiCompact (pullback.snd : pullback f g ⟶ Y) :=
+  quasiCompact_stableUnderBaseChange.snd f g inferInstance
+
+@[elab_as_elim]
+theorem compact_open_induction_on {P : Opens X.carrier → Prop} (S : Opens X.carrier)
+    (hS : IsCompact S.1) (h₁ : P ⊥)
+    (h₂ : ∀ (S : Opens X.carrier) (_ : IsCompact S.1) (U : X.affineOpens), P S → P (S ⊔ U)) :
+    P S := by
+  classical
+  obtain ⟨s, hs, hs'⟩ := (isCompact_open_iff_eq_finset_affine_union S.1).mp ⟨hS, S.2⟩
+  replace hs' : S = iSup fun i : s => (i : Opens X.carrier) := by ext1; simpa using hs'
+  subst hs'
+  apply @Set.Finite.induction_on _ _ _ hs
+  · convert h₁; rw [iSup_eq_bot]; rintro ⟨_, h⟩; exact h.elim
+  · intro x s _ hs h₄
+    have : IsCompact (⨆ i : s, (i : Opens X.carrier)).1 := by
+      refine' ((isCompact_open_iff_eq_finset_affine_union _).mpr _).1; exact ⟨s, hs, by simp⟩
+    convert h₂ _ this x h₄
+    rw [iSup_subtype, sup_comm]
+    conv_rhs => rw [iSup_subtype]
+    exact iSup_insert
+#align algebraic_geometry.compact_open_induction_on AlgebraicGeometry.compact_open_induction_on
+
+theorem exists_pow_mul_eq_zero_of_res_basicOpen_eq_zero_of_isAffineOpen (X : Scheme)
+    {U : Opens X.carrier} (hU : IsAffineOpen U) (x f : X.presheaf.obj (op U))
+    (_ : Scheme.basicOpen X f ≤ U) -- this assumption should be found automatically
+    -- by `restrict_tac`, it is added temporarily so as to make `x |_ X.basicOpen f` work just below
+    (H : x |_ X.basicOpen f = 0) : ∃ n : ℕ, f ^ n * x = 0 := by
```

Comment:
The syntax `x |_ X.basicOpen f` does not work because the assumption that `Scheme.basicOpen X f ≤ U` is not found automatically. The problem is that although the lemma `Scheme.basicOpen_le` is tagged with the attribute `sheaf_restrict`, the tactic `restrict_tac` does not seem to search for the lemmas tagged with this attribute.

Reply:
Tbh, I don't know how tactics and Aesop works, I just copied `aesop_cat`.
```lean
macro (name := restrict_tac') "restrict_tac'" : tactic =>
`(tactic|
  aesop? (options := { introsTransparency? := some .reducible, terminal := true })
  (rule_sets [$(Lean.mkIdent `Restrict):ident]))
```

This solves `X.basicOpen f <= U`, however it doesn't solve what `restrict_tac` solve and
```lean
example (X : Scheme)
    {U V : Opens X} (f : X.presheaf.obj (op U))
    : X.basicOpen f ⊓ V ≤ U := by
```
this cannot be solved by either `restrict_tac` or `restrict_tac'`

Reply:
may be @JLimperg can provide some assistance on this

Reply:
```lean
theorem mem_basicOpen_top' {U : Opens X} (f : X.presheaf.obj (op U)) (x : X.carrier) :
    x ∈ X.basicOpen f ↔ ∃ (m : x ∈ U), IsUnit (X.presheaf.germ (⟨x, m⟩ : U) f) := by
```

This lemma helps `restrict_tac` find the desired proofs if it is tagged `@[simp]`, but this is not a very good lemma, so I gave it a local `simp` attribute whenever `restrict_tac` need this lemma

Reply:
Thanks! Unless a better solution is found, it looks good to me. Could you add a porting note?

Reply:
If this lemma is generally useful for `restrict_tac`, you can add it to Aesop's `simp` set (but not the global `simp` set):

```lean
@[aesop simp (rule_sets := [Restrict])]
theorem mem_basicOpen_top' ...
```

I hope I got my own syntax right...

Reply:
@JLimperg, thanks for the suggestion, but Lean is complaining
```
expected '['
```

Reply:
Ah, it is `@[aesop norm simp (rule_sets [Restrict])]`, I wonder what `norm` means

Reply:
It means that the rule is registered in the 'normalisation phase' of Aesop. Aesop applies rules in three phases, of which this is the first. `simp` rules are actually always run in this phase, so Aesop could infer the phase in this case. See the Aesop [README](https://github.com/JLimperg/aesop) for details.